File: tests/dataset/job/compiler/go/q2.go
//go:build ignore

package main

import (
	"encoding/json"
	"fmt"
	"mochi/runtime/data"
	"reflect"
	"time"
)

func expect(cond bool) {
	if !cond {
		panic("expect failed")
	}
}

func formatDuration(d time.Duration) string {
	switch {
	case d < time.Microsecond:
		return fmt.Sprintf("%dns", d.Nanoseconds())
	case d < time.Millisecond:
		return fmt.Sprintf("%.1fµs", float64(d.Microseconds()))
	case d < time.Second:
		return fmt.Sprintf("%.1fms", float64(d.Milliseconds()))
	default:
		return fmt.Sprintf("%.2fs", d.Seconds())
	}
}

func printTestStart(name string) {
	fmt.Printf("   test %-30s ...", name)
}

func printTestPass(d time.Duration) {
	fmt.Printf(" ok (%s)\n", formatDuration(d))
}

func printTestFail(err error, d time.Duration) {
	fmt.Printf(" fail %v (%s)\n", err, formatDuration(d))
}

func test_Q2_finds_earliest_title_for_German_companies_with_character_keyword() {
	expect(_equal(result, "Der Film"))
}

type Company_nameItem struct {
	Id           int    `json:"id"`
	Country_code string `json:"country_code"`
}

var company_name []Company_nameItem

type KeywordItem struct {
	Id      int    `json:"id"`
	Keyword string `json:"keyword"`
}

var keyword []KeywordItem

type Movie_companiesItem struct {
	Movie_id   int `json:"movie_id"`
	Company_id int `json:"company_id"`
}

var movie_companies []Movie_companiesItem

type Movie_keywordItem struct {
	Movie_id   int `json:"movie_id"`
	Keyword_id int `json:"keyword_id"`
}

var movie_keyword []Movie_keywordItem

type TitleItem struct {
	Id    int    `json:"id"`
	Title string `json:"title"`
}

var title []TitleItem
var titles []string
var result any

func main() {
	failures := 0
	company_name = _cast[[]Company_nameItem]([]Company_nameItem{Company_nameItem{
		Id:           1,
		Country_code: "[de]",
	}, Company_nameItem{
		Id:           2,
		Country_code: "[us]",
	}})
	keyword = _cast[[]KeywordItem]([]KeywordItem{KeywordItem{
		Id:      1,
		Keyword: "character-name-in-title",
	}, KeywordItem{
		Id:      2,
		Keyword: "other",
	}})
	movie_companies = _cast[[]Movie_companiesItem]([]Movie_companiesItem{Movie_companiesItem{
		Movie_id:   100,
		Company_id: 1,
	}, Movie_companiesItem{
		Movie_id:   200,
		Company_id: 2,
	}})
	movie_keyword = _cast[[]Movie_keywordItem]([]Movie_keywordItem{Movie_keywordItem{
		Movie_id:   100,
		Keyword_id: 1,
	}, Movie_keywordItem{
		Movie_id:   200,
		Keyword_id: 2,
	}})
	title = _cast[[]TitleItem]([]TitleItem{TitleItem{
		Id:    100,
		Title: "Der Film",
	}, TitleItem{
		Id:    200,
		Title: "Other Movie",
	}})
	titles = func() []string {
		_res := []string{}
		for _, cn := range company_name {
			for _, mc := range movie_companies {
				if !(mc.Company_id == cn.Id) {
					continue
				}
				for _, t := range title {
					if !(mc.Movie_id == t.Id) {
						continue
					}
					for _, mk := range movie_keyword {
						if !(mk.Movie_id == t.Id) {
							continue
						}
						for _, k := range keyword {
							if !(mk.Keyword_id == k.Id) {
								continue
							}
							if ((cn.Country_code == "[de]") && (k.Keyword == "character-name-in-title")) && (mc.Movie_id == mk.Movie_id) {
								if ((cn.Country_code == "[de]") && (k.Keyword == "character-name-in-title")) && (mc.Movie_id == mk.Movie_id) {
									_res = append(_res, t.Title)
								}
							}
						}
					}
				}
			}
		}
		return _res
	}()
	result = _min(titles)
	func() { b, _ := json.Marshal(result); fmt.Println(string(b)) }()
	{
		printTestStart("Q2 finds earliest title for German companies with character keyword")
		start := time.Now()
		var failed error
		func() {
			defer func() {
				if r := recover(); r != nil {
					failed = fmt.Errorf("%v", r)
				}
			}()
			test_Q2_finds_earliest_title_for_German_companies_with_character_keyword()
		}()
		if failed != nil {
			failures++
			printTestFail(failed, time.Since(start))
		} else {
			printTestPass(time.Since(start))
		}
	}
	if failures > 0 {
		fmt.Printf("\n[FAIL] %d test(s) failed.\n", failures)
	}
}

func _cast[T any](v any) T {
	if tv, ok := v.(T); ok {
		return tv
	}
	var out T
	switch any(out).(type) {
	case int:
		switch vv := v.(type) {
		case int:
			return any(vv).(T)
		case float64:
			return any(int(vv)).(T)
		case float32:
			return any(int(vv)).(T)
		}
	case float64:
		switch vv := v.(type) {
		case int:
			return any(float64(vv)).(T)
		case float64:
			return any(vv).(T)
		case float32:
			return any(float64(vv)).(T)
		}
	case float32:
		switch vv := v.(type) {
		case int:
			return any(float32(vv)).(T)
		case float64:
			return any(float32(vv)).(T)
		case float32:
			return any(vv).(T)
		}
	}
	if m, ok := v.(map[any]any); ok {
		v = _convertMapAny(m)
	}
	data, err := json.Marshal(v)
	if err != nil {
		panic(err)
	}
	if err := json.Unmarshal(data, &out); err != nil {
		panic(err)
	}
	return out
}

func _convertMapAny(m map[any]any) map[string]any {
	out := make(map[string]any, len(m))
	for k, v := range m {
		key := fmt.Sprint(k)
		if sub, ok := v.(map[any]any); ok {
			out[key] = _convertMapAny(sub)
		} else {
			out[key] = v
		}
	}
	return out
}

func _equal(a, b any) bool {
	av := reflect.ValueOf(a)
	bv := reflect.ValueOf(b)
	if av.Kind() == reflect.Slice && bv.Kind() == reflect.Slice {
		if av.Len() != bv.Len() {
			return false
		}
		for i := 0; i < av.Len(); i++ {
			if !_equal(av.Index(i).Interface(), bv.Index(i).Interface()) {
				return false
			}
		}
		return true
	}
	if av.Kind() == reflect.Map && bv.Kind() == reflect.Map {
		if av.Len() != bv.Len() {
			return false
		}
		for _, k := range av.MapKeys() {
			bvVal := bv.MapIndex(k)
			if !bvVal.IsValid() {
				return false
			}
			if !_equal(av.MapIndex(k).Interface(), bvVal.Interface()) {
				return false
			}
		}
		return true
	}
	if (av.Kind() == reflect.Int || av.Kind() == reflect.Int64 || av.Kind() == reflect.Float64) &&
		(bv.Kind() == reflect.Int || bv.Kind() == reflect.Int64 || bv.Kind() == reflect.Float64) {
		return av.Convert(reflect.TypeOf(float64(0))).Float() == bv.Convert(reflect.TypeOf(float64(0))).Float()
	}
	return reflect.DeepEqual(a, b)
}

func _min(v any) any {
	if g, ok := v.(*data.Group); ok {
		v = g.Items
	}
	switch s := v.(type) {
	case []int:
		if len(s) == 0 {
			return 0
		}
		m := s[0]
		for _, n := range s[1:] {
			if n < m {
				m = n
			}
		}
		return m
	case []float64:
		if len(s) == 0 {
			return 0.0
		}
		m := s[0]
		for _, n := range s[1:] {
			if n < m {
				m = n
			}
		}
		return m
	case []string:
		if len(s) == 0 {
			return ""
		}
		m := s[0]
		for _, n := range s[1:] {
			if n < m {
				m = n
			}
		}
		return m
	case []any:
		if len(s) == 0 {
			return 0
		}
		switch s[0].(type) {
		case string:
			m := s[0].(string)
			for _, it := range s[1:] {
				v := it.(string)
				if v < m {
					m = v
				}
			}
			return m
		case int, int64, float64:
			var m float64
			var isFloat bool
			switch n := s[0].(type) {
			case int:
				m = float64(n)
			case int64:
				m = float64(n)
			case float64:
				m = n
				isFloat = true
			}
			for _, it := range s[1:] {
				switch v := it.(type) {
				case int:
					if float64(v) < m {
						m = float64(v)
					}
				case int64:
					if float64(v) < m {
						m = float64(v)
					}
				case float64:
					if v < m {
						m = v
					}
					isFloat = true
				}
			}
			if isFloat {
				return m
			}
			return int(m)
		default:
			panic("min() expects numbers or strings")
		}
	default:
		rv := reflect.ValueOf(v)
		if rv.Kind() == reflect.Slice {
			if rv.Len() == 0 {
				return 0
			}
			m := rv.Index(0).Interface()
			switch m.(type) {
			case int, int64, float64:
				items := make([]any, rv.Len())
				for i := 0; i < rv.Len(); i++ {
					items[i] = rv.Index(i).Interface()
				}
				return _min(items)
			}
		}
		panic("min() expects list or group")
	}
}

File: tests/dataset/job/compiler/c/q2.c

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef struct {
  int len;
  int *data;
} list_int;
static list_int list_int_create(int len) {
  list_int l;
  l.len = len;
  l.data = calloc(len, sizeof(int));
  if (!l.data && len > 0) {
    fprintf(stderr, "alloc failed\n");
    exit(1);
  }
  return l;
}
typedef struct {
  int len;
  double *data;
} list_float;
static list_float list_float_create(int len) {
  list_float l;
  l.len = len;
  l.data = calloc(len, sizeof(double));
  if (!l.data && len > 0) {
    fprintf(stderr, "alloc failed\n");
    exit(1);
  }
  return l;
}
typedef struct {
  int len;
  char **data;
} list_string;
static list_string list_string_create(int len) {
  list_string l;
  l.len = len;
  l.data = calloc(len, sizeof(char *));
  if (!l.data && len > 0) {
    fprintf(stderr, "alloc failed\n");
    exit(1);
  }
  return l;
}
typedef struct {
  int len;
  list_int *data;
} list_list_int;
static list_list_int list_list_int_create(int len) {
  list_list_int l;
  l.len = len;
  l.data = calloc(len, sizeof(list_int));
  if (!l.data && len > 0) {
    fprintf(stderr, "alloc failed\n");
    exit(1);
  }
  return l;
}
static char *_min_string(list_string v) {
  if (v.len == 0)
    return "";
  char *m = v.data[0];
  for (int i = 1; i < v.len; i++)
    if (strcmp(v.data[i], m) < 0)
      m = v.data[i];
  return m;
}
static void _json_int(int v) { printf("%d", v); }
static void _json_float(double v) { printf("%g", v); }
static void _json_string(char *s) { printf("\"%s\"", s); }
static void _json_list_int(list_int v) {
  printf("[");
  for (int i = 0; i < v.len; i++) {
    if (i > 0)
      printf(",");
    _json_int(v.data[i]);
  }
  printf("]");
}
static void _json_list_float(list_float v) {
  printf("[");
  for (int i = 0; i < v.len; i++) {
    if (i > 0)
      printf(",");
    _json_float(v.data[i]);
  }
  printf("]");
}
static void _json_list_string(list_string v) {
  printf("[");
  for (int i = 0; i < v.len; i++) {
    if (i > 0)
      printf(",");
    _json_string(v.data[i]);
  }
  printf("]");
}
static void _json_list_list_int(list_list_int v) {
  printf("[");
  for (int i = 0; i < v.len; i++) {
    if (i > 0)
      printf(",");
    _json_list_int(v.data[i]);
  }
  printf("]");
}
typedef struct {
  int id;
  char *country_code;
} Company_nameItem;
typedef struct {
  int len;
  Company_nameItem *data;
} list_Company_nameItem;
static list_Company_nameItem list_Company_nameItem_create(int len) {
  list_Company_nameItem l;
  l.len = len;
  l.data = calloc(len, sizeof(Company_nameItem));
  if (!l.data && len > 0) {
    fprintf(stderr, "alloc failed\n");
    exit(1);
  }
  return l;
}

typedef struct {
  int id;
  char *keyword;
} KeywordItem;
typedef struct {
  int len;
  KeywordItem *data;
} list_KeywordItem;
static list_KeywordItem list_KeywordItem_create(int len) {
  list_KeywordItem l;
  l.len = len;
  l.data = calloc(len, sizeof(KeywordItem));
  if (!l.data && len > 0) {
    fprintf(stderr, "alloc failed\n");
    exit(1);
  }
  return l;
}

typedef struct {
  int movie_id;
  int company_id;
} Movie_companiesItem;
typedef struct {
  int len;
  Movie_companiesItem *data;
} list_Movie_companiesItem;
static list_Movie_companiesItem list_Movie_companiesItem_create(int len) {
  list_Movie_companiesItem l;
  l.len = len;
  l.data = calloc(len, sizeof(Movie_companiesItem));
  if (!l.data && len > 0) {
    fprintf(stderr, "alloc failed\n");
    exit(1);
  }
  return l;
}

typedef struct {
  int movie_id;
  int keyword_id;
} Movie_keywordItem;
typedef struct {
  int len;
  Movie_keywordItem *data;
} list_Movie_keywordItem;
static list_Movie_keywordItem list_Movie_keywordItem_create(int len) {
  list_Movie_keywordItem l;
  l.len = len;
  l.data = calloc(len, sizeof(Movie_keywordItem));
  if (!l.data && len > 0) {
    fprintf(stderr, "alloc failed\n");
    exit(1);
  }
  return l;
}

typedef struct {
  int id;
  char *title;
} TitleItem;
typedef struct {
  int len;
  TitleItem *data;
} list_TitleItem;
static list_TitleItem list_TitleItem_create(int len) {
  list_TitleItem l;
  l.len = len;
  l.data = calloc(len, sizeof(TitleItem));
  if (!l.data && len > 0) {
    fprintf(stderr, "alloc failed\n");
    exit(1);
  }
  return l;
}

static int
    test_Q2_finds_earliest_title_for_German_companies_with_character_keyword_result;
static void
test_Q2_finds_earliest_title_for_German_companies_with_character_keyword() {
  if (!((test_Q2_finds_earliest_title_for_German_companies_with_character_keyword_result ==
         "Der Film"))) {
    fprintf(stderr, "expect failed\n");
    exit(1);
  }
}

int main() {
  Company_nameItem tmp1_data[] = {
      (Company_nameItem){.id = 1, .country_code = "[de]"},
      (Company_nameItem){.id = 2, .country_code = "[us]"}};
  list_Company_nameItem tmp1 = {2, tmp1_data};
  list_Company_nameItem company_name = tmp1;
  KeywordItem tmp2_data[] = {
      (KeywordItem){.id = 1, .keyword = "character-name-in-title"},
      (KeywordItem){.id = 2, .keyword = "other"}};
  list_KeywordItem tmp2 = {2, tmp2_data};
  list_KeywordItem keyword = tmp2;
  Movie_companiesItem tmp3_data[] = {
      (Movie_companiesItem){.movie_id = 100, .company_id = 1},
      (Movie_companiesItem){.movie_id = 200, .company_id = 2}};
  list_Movie_companiesItem tmp3 = {2, tmp3_data};
  list_Movie_companiesItem movie_companies = tmp3;
  Movie_keywordItem tmp4_data[] = {
      (Movie_keywordItem){.movie_id = 100, .keyword_id = 1},
      (Movie_keywordItem){.movie_id = 200, .keyword_id = 2}};
  list_Movie_keywordItem tmp4 = {2, tmp4_data};
  list_Movie_keywordItem movie_keyword = tmp4;
  TitleItem tmp5_data[] = {(TitleItem){.id = 100, .title = "Der Film"},
                           (TitleItem){.id = 200, .title = "Other Movie"}};
  list_TitleItem tmp5 = {2, tmp5_data};
  list_TitleItem title = tmp5;
  list_string tmp6 =
      list_string_create(company_name.len * movie_companies.len * title.len *
                         movie_keyword.len * keyword.len);
  int tmp7 = 0;
  for (int tmp8 = 0; tmp8 < company_name.len; tmp8++) {
    Company_nameItem cn = company_name.data[tmp8];
    for (int tmp9 = 0; tmp9 < movie_companies.len; tmp9++) {
      Movie_companiesItem mc = movie_companies.data[tmp9];
      if (!(mc.company_id == cn.id)) {
        continue;
      }
      for (int tmp10 = 0; tmp10 < title.len; tmp10++) {
        TitleItem t = title.data[tmp10];
        if (!(mc.movie_id == t.id)) {
          continue;
        }
        for (int tmp11 = 0; tmp11 < movie_keyword.len; tmp11++) {
          Movie_keywordItem mk = movie_keyword.data[tmp11];
          if (!(mk.movie_id == t.id)) {
            continue;
          }
          for (int tmp12 = 0; tmp12 < keyword.len; tmp12++) {
            KeywordItem k = keyword.data[tmp12];
            if (!(mk.keyword_id == k.id)) {
              continue;
            }
            if (!((strcmp(cn.country_code, "[de]") == 0) &&
                  k.keyword == "character-name-in-title" &&
                  mc.movie_id == mk.movie_id)) {
              continue;
            }
            tmp6.data[tmp7] = t.title;
            tmp7++;
          }
        }
      }
    }
  }
  tmp6.len = tmp7;
  list_string titles = tmp6;
  int result = _min_string(titles);
  _json_int(result);
  test_Q2_finds_earliest_title_for_German_companies_with_character_keyword_result =
      result;
  test_Q2_finds_earliest_title_for_German_companies_with_character_keyword();
  return 0;
}
